projects/cli-files/go-cat/cmd/root.go

```diff
@@ -0,0 +1,66 @@
+package cmd
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+)
+
+func Execute() {
+	args := os.Args[1:]
+
+	for _, arg := range args {
+		err := checkArgs(arg)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return
+		}
+		fileLines, err := readFileLines(arg)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		for _, fileLine := range fileLines {
+			fmt.Print(fileLine)
+		}
+	}
+}
+
+func checkArgs(arg string) error {
+	if arg == "" {
+		return fmt.Errorf("go-cat: no file specified")
+	}
+	fileInfo, err := os.Stat(arg)
+	if err != nil {
+		return fmt.Errorf("go-cat: '%s': no such file or directory", arg)
+	}
+	if fileInfo.IsDir() {
+		return fmt.Errorf("go-cat: '%s': is a directory", arg)
+	}
+	return nil
+}
+
+func readFileLines(filePath string) ([]string, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	reader := bufio.NewReader(file)
+	buffer := make([]byte, 16)
+	fileLines := []string{}
+
+	for {
+		n, err := reader.Read(buffer)
+		if err != nil {
+			if err != io.EOF {
+				return nil, err
+			}
+			fileLines = append(fileLines, "\n")
```

Comment:
From [the documentation of `io.Reader`](https://pkg.go.dev/io#Reader):

> When Read encounters an error or end-of-file condition after successfully reading n > 0 bytes, it returns the number of bytes read. It may return the (non-nil) error from the same call or return the error (and n == 0) from a subsequent call.

Which means we actually need to handle the idea that _both_ `err = io.EOF` _and_ `buffer` contains some bytes we need to append.

Also, it's possible the file _doesn't_ end in a newline, so it's a little odd to unconditionally add a newline at the end?

Reply:
can you explain this further please. I'm not sure I get your point.

Reply:
Imagine we're reading a file with contents "123", and we're reading one byte at a time. It's valid for either of these to be the set of written bytes and return values we observe for each call to `Read`:

1. b"1", 1, nil
2. b"2",1 , nil
3. b"3", 1, nil
4. b"", 0, io.EOF

or

1. b"1", 1, nil
2. b"2", 1, nil
3. b"3", 1, io.EOF

Your code will currently only handle the first case - you're currently assuming that if `io.EOF` was returned, no bytes were written to the slice.

If the reader happened to return values like the second way (where at some point it says both "I read one byte" _and_ "this is the end of the file), you would miss the last byte in what you accumulate.

(Note: Personally I think this is bad API design in the `io.Reader` interface which makes it unnecessarily complicated to use, but that's a discussion for another day :))

Reply:
So by adding a condition like `n==0` is more relevant than the `err == io.EOF` before break. It will handle both cases.

Reply:
I think there are two _separate_ conditions you need:

You need to guard whether you're going to append bytes on `n==0`

And separately you need to guard whether you're going to break on `err == io.EOF`

Reply:
> You need to guard whether you're going to append bytes on n==0
if there is no bytes to read why do I need to append bytes ?

isn't `n == 0 && err == io.EOF ` enough to break the loop and make sure that we're reading the whole file?

Reply:
I think it's more clear to model these as separate concerns:

```go
if n != 0 {
    fileLines = append(fileLines, string(buffer[0:n]))
}
if err == io.EOF {
    break
}
```

Whether there _are_ bytes to read, and whether you need to read bytes again, are separate concepts, so treating them separately is more simple and clear than coupling them together.

Otherwise e.g. in the second example I gave above, you will do a 4th read which will return `0, io.EOF`, rather than stopping after 3.

Reply:
I see, I was focusing on the `n==0` and forgot to think of the opposite case that we can control. :)